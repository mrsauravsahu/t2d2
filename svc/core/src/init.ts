import { execa, ExecaChildProcess } from 'execa'

import { InitOpts } from './shapes/init-options';
import { Profile } from './shapes/profile';

export const init = async (options: InitOpts) => {
  const { stdout, stderr } = await execa('terraform', ['init'], {
    cwd: options.workspaceDir
  })

  stdout !== '' ? console.log(stdout) : undefined
  stderr!== '' ? console.log(stderr) : undefined

  const profile: Profile = {
    name: options.profileName,
    workspaceDir: options.workspaceDir
  }

  return profile;
}
