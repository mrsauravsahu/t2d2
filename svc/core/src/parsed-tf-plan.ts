import { TFResource } from "./shapes/resource";
import { TFModule } from "./shapes/module";
import { IParsedTFPlan } from "./shapes/i-parsed-tf-plan";

export class ParsedTFPlan {
  constructor(private parsedPlan: IParsedTFPlan) { }

  public get resources(): TFResource[] {
    return [
      ...this.parsedPlan.rootModuleResources,
      ...this.parsedPlan.childModules.map(m => m.resources).flat()
    ]
  }

  public get rawState() { return this.parsedPlan.rawTFStateJsonString; }
  public get formatVersion(): string { return this.parsedPlan.formatVersion }
  public get tfVersion(): string { return this.parsedPlan.terraformVersion }

  public getResourceByAddress(address: string): TFResource {
    const matchedResources = this.resources.filter(r => r.address === address)
    if (matchedResources.length === 0)
      throw new Error(`Resource with address '${address}' does not exist.`)
    return matchedResources[0]
  }

  public getRootModuleResourceOfType(type: string): TFResource {
    const matchedResources = this.parsedPlan.rootModuleResources.filter(r => r.type === type)
    if (matchedResources.length === 0)
      throw new Error(`Resource of type '${type}' does not exist in the root module.`)
    return matchedResources[0]
  }

  public getModuleByAddress(address: string): TFModule {
    const matchedModules = this.parsedPlan.childModules.filter(m => m.address === address)
    if (matchedModules.length === 0) throw `Module at address '${address}' does not exist.`
    return  matchedModules[0]
  }
}