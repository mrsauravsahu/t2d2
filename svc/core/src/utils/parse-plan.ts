import * as jq from 'node-jq'

import { ParsedTFPlan } from "../parsed-tf-plan";
import { TFModule } from '../shapes/module';
import { TFResource } from '../shapes/resource';

// TODO: parse differently based on format_version
export const parsePlan = async (rawTFStateJsonString: string): Promise<ParsedTFPlan> => {
    const jqOptions = { input: 'string', output: 'json' }

    const versionData = await jq.run(
        `{
            formatVersion: .format_version,
            terraformVersion: .terraform_version
        }`,
        rawTFStateJsonString,
        jqOptions
    );

    const rootModuleResources: unknown = await jq.run(
        '[(.planned_values.root_module.resources // [])[]]',
        rawTFStateJsonString,
        jqOptions);

    const modules: unknown = await jq.run(
        '[(.planned_values.root_module.child_modules // [])[]]',
        rawTFStateJsonString,
        jqOptions);

    return new ParsedTFPlan(
        (versionData as any).formatVersion,
        (versionData as any).terraformVersion,
        rootModuleResources as unknown as TFResource[],
        modules as unknown as TFModule[],
        // TODO
        new Map<string, { value: string }>(),
        rawTFStateJsonString
    );
}
