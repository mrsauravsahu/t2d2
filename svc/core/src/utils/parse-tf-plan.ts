import * as jq from 'node-jq'

import { IParsedTFPlan } from "../shapes/i-parsed-tf-plan";
import { TFModule } from '../shapes/module';
import { TFResource } from '../shapes/resource';

// TODO: parse differently based on format_version
export const parseTFPlan = async (rawTFStateJsonString: string): Promise<IParsedTFPlan> => {
    const jqOptions = { input: 'string', output: 'json' }

    const versionData = await jq.run(
        `{
            formatVersion: .format_version,
            terraformVersion: .terraform_version
        }`,
        rawTFStateJsonString,
        jqOptions
    );

    const rootModuleResources: unknown = await jq.run(
        '[(.planned_values.root_module.resources // [])[]]',
        rawTFStateJsonString,
        jqOptions);

    const childModules: unknown = await jq.run(
        '[(.planned_values.root_module.child_modules // [])[]]',
        rawTFStateJsonString,
        jqOptions);

    return {
        formatVersion: (versionData as any).formatVersion,
        terraformVersion: (versionData as any).terraformVersion,
        rootModuleResources: rootModuleResources as unknown as TFResource[],
        childModules: childModules as unknown as TFModule[],
        // TODO
        variables: new Map<string, { value: string }>(),
        rawTFStateJsonString
    }
}
