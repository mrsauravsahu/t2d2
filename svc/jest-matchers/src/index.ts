export * from './register-jest-matchers'

declare global {
  namespace jest {
    interface Matchers<R> {
      toHaveRootModuleResourceOfType(type: string): jest.CustomMatcherResult;
    }
  }
}
