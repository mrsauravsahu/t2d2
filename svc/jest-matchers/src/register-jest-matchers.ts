import { ParsedTFPlan } from "@t2d2/core";

const toHaveRootModuleResourceOfType = (plan: ParsedTFPlan, type: string): jest.CustomMatcherResult => {
  try {
    plan.getRootModuleResourceOfType(type)
    return { pass: true, message: () => `The plan does contain a resource of type '${type}' at the root module.` }
  }
  catch (error) {
    return { pass: false, message: () => (error as Error).message }
  }
}

const toHaveModuleWithAddress = (plan: ParsedTFPlan, address: string): jest.CustomMatcherResult => {
  try {
    plan.getModuleByAddress(address)
    return { pass: true, message: () => `The plan does contains a module with address '${address}'.` }
  }
  catch (error) {
    return { pass: false, message: () => (error as Error).message }
  }
}

const toHaveResourceWithAddress = (plan: ParsedTFPlan, address: string): jest.CustomMatcherResult => {
  try {
    plan.getResourceByAddress(address)
    return { pass: true, message: () => `The plan does contains a resource at address '${address}'.` }
  }
  catch (error) {
    return { pass: false, message: () => (error as Error).message }
  }
}


export const registerJestMatchers = () => {
  expect.extend({ toHaveRootModuleResourceOfType });
  expect.extend({ toHaveModuleWithAddress });
  expect.extend({ toHaveResourceWithAddress });
}
