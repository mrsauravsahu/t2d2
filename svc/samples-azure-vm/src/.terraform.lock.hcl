# This file is maintained automatically by "terraform init".
# Manual edits may be lost in future updates.

provider "registry.terraform.io/hashicorp/azurerm" {
  version     = "2.98.0"
  constraints = "2.98.0"
  hashes = [
    "h1:5AmkI94p8OwglrqFHD6rt6jZFqVSKHavBBCn6FRdipk=",
    "zh:025f656a6d3ecc30f7cc2279bc41969789987b405e3fa8a7c1eb5f74e3ee1140",
    "zh:23c54b330678a16378156193d709bbddce3ba76ee827fd65fb751ce90790af9e",
    "zh:2d28d359ce6881918bd6c03701f6ec4fd90215abfce9b863cfd3172e28c1acb3",
    "zh:31df88584d39cf876fa45ff6de92e67e03814a0985d34c7671bd6989cda22af8",
    "zh:36019109790b9a905770355e5bbb57b291a9689a8b9beac5751dcbdb1282d035",
    "zh:5fb4a277331c459db9e1b150d79b7c7157a176ceca871195e81225e949141b72",
    "zh:7ec304afa1b60dc84257a54cea68e97f85df3feb405d25a9226a4f593ed00744",
    "zh:bac469f104b8ad2c8b5ddc88ddae3b0bc27ae5f9c2ccf03f14a001a5c3ed6ae1",
    "zh:d860b0ec60a978fe3f08d695326e9051a61cd3f60786fc618a61fbdb5d6a4f15",
    "zh:ebcb2911ee27587f63df7eff3836c9a206181a931357c6b9a380124be4241597",
    "zh:f37fae57bf7d05c30fda6e5414ad5a4aad1b34d41a5f2465a864736f92ded1ac",
  ]
}
