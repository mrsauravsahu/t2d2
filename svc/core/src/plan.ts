import { writeFile } from 'fs/promises';
import * as path from 'path';
import { execa } from 'execa';
import { PlanOptions } from './shapes/plan-options';
import { Profile } from './shapes/profile';
import { ParsedTFPlan } from './parsed-tf-plan';
import { parseTFPlan } from './utils/parse-tf-plan';

export const plan = async (profile: Profile, options: PlanOptions): Promise<ParsedTFPlan> => {
  var planArgs: string[] = [
    'plan',
    ...(options.varFile ? ['-var-file', options.varFile] : []),
    ...(options.vars ? Object.keys(options.vars).map((key) => ['-var', `${key}=${(options?.vars || {})[key] ?? ''}`]) : []).flat(),
    "-out=terraform.tfstate.tmp"
  ]

  const plan = await execa('terraform', planArgs, {
    cwd: profile.workspaceDir
  });

  plan.stdout !== '' ? console.log(plan.stdout) : undefined
  plan.stderr !== '' ? console.log(plan.stderr) : undefined

  var showArgs: string[] = [
    "show",
    "-json",
    "terraform.tfstate.tmp"
  ]

  const show = await execa('terraform', showArgs, {
    cwd: profile.workspaceDir
  });

  show.stdout !== '' ? console.log(show.stdout) : undefined
  show.stderr !== '' ? console.log(show.stderr) : undefined

  const tfPlanJsonPath = path.join(profile.workspaceDir, 'plan.json');
  await writeFile(tfPlanJsonPath, show.stdout, { encoding: 'utf-8' })

  const parsedPlanContract = await parseTFPlan(show.stdout)
  const parsedPlan = new ParsedTFPlan(parsedPlanContract)
  return parsedPlan
}
