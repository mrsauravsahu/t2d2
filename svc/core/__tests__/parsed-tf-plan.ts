import * as fs from 'fs/promises'
import * as path from 'path'

import { ParsedTFPlan } from "../src"
import { parseTFPlan } from "../src/utils/parse-tf-plan"

const loadParsedPlanFromRelativeInputFile = async (inputFile: string): Promise<ParsedTFPlan> => {
  const parsedPlan = await parseTFPlan(await fs.readFile(
    path.resolve(__dirname, `./inputs/${inputFile}`),
    { encoding: 'utf-8' }
  ))
  const plan = new ParsedTFPlan(parsedPlan)
  return plan
}

describe('ParsedTFPlan', () => {
  describe('getResourceByAddress', () => {
    let plan: ParsedTFPlan;
    beforeAll(async () => { plan = await loadParsedPlanFromRelativeInputFile('basic-without-modules.input.json') })

    test('should throw Error if the required resource does not exist', () => {
      expect(() => { plan.getResourceByAddress('not_present.not_present') })
        .toThrowError('Resource with address \'not_present.not_present\' does not exist.')
    })

    test('should match resource snapshot if exists', () => {
      expect(plan.getResourceByAddress('time_sleep.wait_some_time')).toMatchSnapshot()
    })
  })

  describe('getRootModuleResourceOfType', () => {
    let plan: ParsedTFPlan;
    beforeAll(async () => { plan = await loadParsedPlanFromRelativeInputFile('basic-without-modules.input.json') })

    test('should throw Error if the required root resource of type does not exist', () => {
      expect(() => { plan.getRootModuleResourceOfType('not_existent_resource_type') })
        .toThrowError('Resource of type \'not_existent_resource_type\' does not exist in the root module.')
    })

    test('should match snapshot if resource of type exists', () => {
      expect(plan.getRootModuleResourceOfType('time_sleep')).toMatchSnapshot()
    })
  })

  describe('getModuleByAddress', () => {
    let plan: ParsedTFPlan;
    beforeAll(async () => {
      plan = await loadParsedPlanFromRelativeInputFile('get-module-by-address.input.json')
    })

    test('should throw Error if module does not exist', () => {
      expect(() => { plan.getModuleByAddress('module.not_present') })
        .toThrowError('Module at address \'module.not_present\' does not exist.')
    })

    test('should match module snapshot if exists', () => {
      expect(plan.getModuleByAddress('module.files')).toMatchSnapshot();
    })
  })
})
