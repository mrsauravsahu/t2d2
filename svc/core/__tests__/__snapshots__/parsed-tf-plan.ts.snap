// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParsedTFPlan modules getModuleByAddress should match module snapshot if exists 1`] = `
Object {
  "address": "module.files",
  "resources": Array [
    Object {
      "address": "module.files.local_file.sample",
      "mode": "managed",
      "name": "sample",
      "provider_name": "registry.terraform.io/hashicorp/local",
      "schema_version": 0,
      "type": "local_file",
      "values": Object {
        "content": "0 1 1 2 3 5 8 13 21",
        "content_base64": null,
        "directory_permission": "0777",
        "file_permission": "0777",
        "filename": "output.txt",
        "sensitive_content": null,
        "source": null,
      },
    },
  ],
}
`;
